src/local/cost_matrix.rs

```diff
@@ -265,12 +302,130 @@ impl IndexMut<Position> for LocalCostMatrix {
 //     }
 // }
 
+#[derive(Clone, Debug)]
+pub struct SparseCostMatrix {
+    inner: HashMap<(u8, u8), u8>
+}
+
+impl Default for SparseCostMatrix {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SparseCostMatrix {
+    pub fn new() -> Self {
+        SparseCostMatrix { inner: HashMap::new() }
+    }
+
+    pub fn get(&self, x: u8, y: u8) -> u8 {
+        assert!(x < 50, "out of bounds x: {}", x);
+        assert!(y < 50, "out of bounds y: {}", y);
+        if let Some(ref_val) = self.inner.get(&(x, y)) {
+            *ref_val
+        } else {
+            0
+        }
+    }
+
+    pub fn set(&mut self, x: u8, y: u8, val: u8) {
+        assert!(x < 50, "out of bounds x: {}", x);
+        assert!(y < 50, "out of bounds y: {}", y);
+        self.inner.insert((x, y), val);
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = ((u8, u8), &u8)> {
+        self.inner.iter().map(|(pos, val)| { (*pos, val) })
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = ((u8, u8), &mut u8)> {
+        self.inner.iter_mut().map(|(pos, val)| { (*pos, val) })
+    }
+
+    // Takes all non-zero entries in `src`, and inserts them into `self`.
+    //
+    // If an entry for that position exists already, overwrites it with the new
+    // value.
+    pub fn merge_from_dense(&mut self, src: &LocalCostMatrix) {
+        self.inner.extend(src.iter().filter_map(|(xy, val)| {
+            if *val > 0 {
+                Some((xy, *val))
+            } else {
+                None
+            }
+        }))
+    }
+
+    // Takes all entries in `src` and merges them into `self`.
+    //
+    // If an entry for that position exists already, overwrites it with the new
+    // value.
+    pub fn merge_from_sparse(&mut self, src: &SparseCostMatrix) {
+        self.inner.extend(src.inner.iter());
+    }
+}
+
+impl From<HashMap<(u8, u8), u8>> for SparseCostMatrix {
+    fn from(mut map: HashMap<(u8, u8), u8>) -> Self {
+        map.retain(|pos, _| { pos.0 < 50 && pos.1 < 50 });
+        SparseCostMatrix { inner: map }
+    }
+}
+
+impl From<HashMap<Position, u8>> for SparseCostMatrix {
+    fn from(mut map: HashMap<Position, u8>) -> Self {
+        SparseCostMatrix { inner: map.drain().map(|(pos, val)| { (pos.into(), val) }).collect() }
+    }
+}
+
+impl From<CostMatrix> for SparseCostMatrix {
+    fn from(js_matrix: CostMatrix) -> Self {
+        let vals: Vec<u8> = js_matrix.get_bits().to_vec();
+        assert!(vals.len() == 2500, "JS CostMatrix had length {} instead of 2500.", vals.len());
+
+        SparseCostMatrix {
+            inner: vals.into_iter().enumerate().filter_map(|(idx, val)| {
+                    if val > 0 {
```

Comment:
Might be worth a small comment here that '0' is treated as default cost by the screeps code so that's why it's filtered out here. (It's equivalent to unset.)